### ISSUE TITLE
New Rule: require global regular expression to be used with `replaceAll`

### ISSUE THREAD

**Issue by Tyler-Murphy** [eslint/eslint#15073]
### Rule details

When `replaceAll` is given a regular expression as its first argument, that regular expression must be global. If it's not global, a runtime error is thrown. This rule should require that regular expressions passed as the first argument to `replaceAll` are global.

### Related ECMAScript feature

`String.prototype.replaceAll`: https://github.com/tc39/proposal-string-replaceall

### What type of rule is this?

Warns about a potential problem

### Example code

```js
`abc`.replaceAll(/a/, ``) // throws Uncaught TypeError: String.prototype.replaceAll called with a non-global RegExp argument

// Should be an error, and could also have an auto-fix to change the code to `abc`.replaceAll(/a/g, ``)
```

### Why should this rule be in the core instead of a plugin?

It prevents a runtime error that's not otherwise apparent.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

It'd be nice to reopen https://github.com/eslint/eslint/issues/12557, too.

**Comment by nzakas** [comment]
Thanks for the suggestion. We don't normally do parameter checks for methods like this because there's a never-ending number of builtin JS methods that expect certain inputs. In this case, if the first argument isn't a regex literal or a `RegExp` object, then we wouldn't be able to warn at all.

My opinion is that this isn't well-suited to be a core rule, but you can always create a custom rule to do this.

**Comment by Tyler-Murphy** [comment]
That's a good point. This strikes me as a little different because most other situations that can cause runtime errors in the JS standard library can be caught by compile-time type checking.

Normally, I'd rely on Typescript to catch an error like this, but I don't think it has a way to represent regular expressions with certain flags. It'd probably be best to pursue adding it there.

It sounds like `no-restricted-syntax` might be able to cover the case I'm thinking of. Do you think it could?

**Comment by mdjermanovic** [comment]
> It sounds like `no-restricted-syntax` might be able to cover the case I'm thinking of. Do you think it could?

This selector would cover regex literals:

```css
CallExpression[callee.property.name='replaceAll'] > Literal[regex.flags=/^[^g]*$/].arguments:first-child
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICBzZWxlY3RvcjogXCJDYWxsRXhwcmVzc2lvbltjYWxsZWUucHJvcGVydHkubmFtZT0ncmVwbGFjZUFsbCddID4gTGl0ZXJhbFtyZWdleC5mbGFncz0vXlteZ10qJC9dLmFyZ3VtZW50czpmaXJzdC1jaGlsZFwiLFxuICAgbWVzc2FnZTogXCJNaXNzaW5nICdnJyBmbGFnLlwiXG59XSovXG5cbmBhYmNgLnJlcGxhY2VBbGwoL2EvLCBgYCk7IC8vIGVycm9yXG5cbmBhYmNgLnJlcGxhY2VBbGwoL2EvdSwgYGApOyAvLyBlcnJvclxuXG5gYWJjYC5yZXBsYWNlQWxsKC9hL2csIGBgKTsgLy8gb2tcblxuYGFiY2AucmVwbGFjZUFsbCgvYS9ndSwgYGApOyAvLyBva1xuXG5gYWJjYC5yZXBsYWNlQWxsKFwiYVwiLCBgYCk7IC8vIG9rIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Comment by bmish** [comment]
There is a third-party rule for this: [unicorn/prefer-string-replace-all](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-string-replace-all.md).

**Comment by Tyler-Murphy** [comment]
@bmish, I think that's different. Maybe for https://github.com/eslint/eslint/issues/12557?
